commit a33999fed4c09c00c2f08cda91603b812bed5271 (refs/changes/82/5282/5)
Author: Rui Loura <rui@joyent.com>
Date:   2018-12-21T19:15:06+00:00 (10 months ago)
    
    TRITON-1058 cnapi server uuid validation should handle "default"
    Reviewed by: Trent Mick <trentm@gmail.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/lib/endpoints/index.js b/lib/endpoints/index.js
index 318b691..072d155 100644
--- a/lib/endpoints/index.js
+++ b/lib/endpoints/index.js
@@ -119,13 +119,15 @@ function prepopulateServer(opts, req, res, next) {
     assert.object(req, 'req');
     assert.object(req.params, 'req.params');
 
-    if (!isUuid(req.params.server_uuid)) {
+    var server_uuid = req.params.server_uuid;
+
+    if (server_uuid !== 'default' && !isUuid(server_uuid)) {
         next(new restify.InvalidArgumentError('Invalid server_uuid'));
         return;
     }
 
     ModelServer.get(
-        req.params.server_uuid,
+        server_uuid,
         function (err, servermodel, server) {
             if (err) {
                 req.log.error(err);
@@ -135,8 +137,7 @@ function prepopulateServer(opts, req, res, next) {
 
             // Check if any servers were returned
             if (!server) {
-                var errorMsg = 'Server ' + req.params.server_uuid
-                    + ' not found';
+                var errorMsg = 'Server ' + server_uuid + ' not found';
                 next(new restify.ResourceNotFoundError(errorMsg));
                 return;
             }
diff --git a/lib/models/server.js b/lib/models/server.js
index 7845fa0..501a723 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -70,6 +70,8 @@ function ModelServer(uuid) {
     }
 
     this.value = null;
+
+    // This can be a server UUID or the special 'default' value.
     this.uuid = uuid;
 
     this.log = ModelServer.getLog();
diff --git a/package.json b/package.json
index 5a87db4..dbdddf4 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.18.1",
+  "version": "1.18.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
diff --git a/test/test-servers.js b/test/test-servers.js
index ca72795..c45cd11 100644
--- a/test/test-servers.js
+++ b/test/test-servers.js
@@ -171,6 +171,17 @@ function testGetServer(t) {
 }
 
 
+function testGetDefaultServer(t) {
+    client.get('/servers/default', function (err, req, res, body) {
+        t.ifError(err);
+        t.equal(res.statusCode, '200',
+            'expecting default server with status code 200');
+        t.equal(body.uuid, 'default', 'default server uuid is "default"');
+        t.done();
+    });
+}
+
+
 function testUpdateServer(t) {
     var uuid;
     var oldRatio;
@@ -794,6 +805,7 @@ module.exports = {
     'list servers with all 2': testListServersWithAll2,
     'list servers using unknown parameter': testListServersUnknownParam,
     'get server': testGetServer,
+    'get default server': testGetDefaultServer,
     'update server': testUpdateServer,
     'update server overprovision ratios': testUpdateServerOverprovisionRatios
 };
